### PR TITLE
WPB-10658 invitation and acceptance of individual users to teams

### PR DESCRIPTION
**Pull request by battermann**
https://wearezeta.atlassian.net/browse/WPB-10658

## Checklist

 - [x] Add a new entry in an appropriate subdirectory of `changelog.d`
 - [x] Read and follow the [PR guidelines](https://docs.wire.com/developer/developer/pr-guidelines.html)
 - [x] Use the same email templates as for normal team invitations. These will be replaced in a subsequent PR.
 - [x] Update all relevant environments before merging (staging, prod?)
 - [x] Add password protection on accepting the invitation

### CODE REVIEW

**Review by mdimjasevic — APPROVED** [review]
There are still a few small things to do, but this looks good.

File: charts/brig/templates/configmap.yaml [hunk 1]
```diff
@@ -179,14 +179,17 @@ data:
...
       {{- if .emailSMS.team }}
         tInvitationUrl: {{ .emailSMS.team.tInvitationUrl }}
+        tExistingUserInvitationUrl: {{ .emailSMS.team.tExistingUserInvitationUrl }}
         tActivationUrl: {{ .emailSMS.team.tActivationUrl }}
         tCreatorWelcomeUrl: {{ .emailSMS.team.tCreatorWelcomeUrl }}
         tMemberWelcomeUrl: {{ .emailSMS.team.tMemberWelcomeUrl }}
       {{- else }}
       {{- if .externalUrls.teamSettings }}
         tInvitationUrl: {{ .externalUrls.teamSettings }}/join/?team-code=${code}
+        tExistingUserInvitationUrl: {{ .externalUrls.teamSettings }}/accept-invitation/?team-code=${code}
```
Comment (fisx) [inline charts/brig/templates/configmap.yaml:189]:
```suggestion
        tExistingUserInvitationUrl: {{ .externalUrls.teamSettings }}/accept-invitation/?team_code=${code}
```

File: charts/brig/templates/configmap.yaml [hunk 2]
```diff
@@ -179,14 +179,17 @@ data:
...
       {{- else }}
         tInvitationUrl: {{ .externalUrls.nginz }}/register?team=${team}&team_code=${code}
+        tExistingUserInvitationUrl: {{ .externalUrls.nginz }}/accept-invitation/?team-code=${code}
```
Comment (fisx) [inline charts/brig/templates/configmap.yaml:192]:
```suggestion
        tExistingUserInvitationUrl: {{ .externalUrls.nginz }}/accept-invitation/?team_code=${code}
```

File: integration/test/Test/Teams.hs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  runCodensity (acquireResources 1 resourcePool) $ \[testBackend] -> do
+    let domain = testBackend.berDomain
+    (owner, tid, tm) <- runCodensity (startDynamicBackend testBackend def) $ \_ -> do
+      (owner, tid, tm : _) <- createTeam domain 2
+      pure (owner, tid, tm)
+
+    runCodensity
+      ( startDynamicBackend
+          testBackend
+          (def {galleyCfg = setField "settings.exposeInvitationURLsTeamAllowlist" [tid]})
```
Comment (fisx) [inline integration/test/Test/Teams.hs:47]: we've done this a few times in the past, and usually within a few months somebody needed this to go from yaml to cql...  i think we should make this a table right away.

File: integration/test/Test/Teams.hs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+          ids `shouldContain` [ownerId]
+        refreshIndex domain
+        -- a team member can now search for the former personal user
+        bindResponse (searchContacts tm (user %. "name") domain) $ \resp -> do
+          resp.status `shouldMatchInt` 200
+          document <- resp.json %. "documents" >>= asList >>= assertOne
+          document %. "id" `shouldMatch` uid
+          document %. "team" `shouldMatch` tid
+
+getQueryParam :: String -> String -> Maybe (Maybe String)
```
Comment (fisx) [inline integration/test/Test/Teams.hs:95]: move this to Testlib.HTTP?

File: libs/wire-api/test/golden/Test/Wire/API/Golden/Manual/UserEvent.hs [hunk 5]
```diff
@@ -99,6 +99,7 @@ testObject_UserEvent_6 =
             Nothing
             False
             (Just mempty)
+            Nothing
```
Comment (fisx) [inline libs/wire-api/test/golden/Test/Wire/API/Golden/Manual/UserEvent.hs:102]: maybe add one with something?

File: services/brig/src/Brig/Team/API.hs [hunk 6]
```diff
@@ -219,15 +232,28 @@ createInvitation' ::
...
-createInvitation' tid mUid inviteeRole mbInviterUid fromEmail body = do
-  let email = (inviteeEmail body)
+createInvitation' tid mUid inviteeRole mbInviterUid fromEmail invRequest = do
+  let email = inviteeEmail invRequest
   let uke = mkEmailKey email
   blacklistedEm <- lift $ liftSem $ isBlocked email
   when blacklistedEm $
     throwStd blacklistedEmail
   emailTaken <- lift $ liftSem $ isJust <$> lookupKey uke
+  migratePersonalUser <-
```
Comment (fisx) [inline services/brig/src/Brig/Team/API.hs:242]:
```suggestion
  weCanMigratePersonalUser <-
```

File: services/brig/src/Brig/Team/API.hs [hunk 7]
```diff
@@ -219,15 +232,28 @@ createInvitation' ::
...
+          -- this can e.g. happen if the key is claimed but the account is not yet created
+          Nothing -> False
+          Just account ->
+            account.accountStatus == Active
+              && isNothing account.accountUser.userTeam
+      else pure False
+
   when emailTaken $
-    throwStd emailExists
+    unless migratePersonalUser $
```
Comment (fisx) [inline services/brig/src/Brig/Team/API.hs:255]: if you define `migratePersonalUser` inside the `when emailTaken` block, you can drop the `if emailTaken` wrapping.

File: services/brig/src/Brig/Team/API.hs [hunk 8]
```diff
@@ -239,20 +265,33 @@ createInvitation' tid mUid inviteeRole mbInviterUid fromEmail body = do
   lift $ do
     iid <- maybe (liftIO DB.mkInvitationId) (pure . Id . toUUID) mUid
     now <- liftIO =<< view currentTime
-    timeout <- setTeamInvitationTimeout <$> view settings
-    (newInv, code) <-
+    timeout <-
+      if migratePersonalUser
+        then pure $ Timeout $ nominalDay * 2 -- todo(Leif): read from settings?
```
Comment (fisx) [inline services/brig/src/Brig/Team/API.hs:270]: yes! or just take `setTeamInvitationTimeout` no matter whats in `migratePersonalUser`.

File: services/brig/src/Brig/Team/API.hs [hunk 9]
```diff
@@ -239,20 +265,33 @@ createInvitation' tid mUid inviteeRole mbInviterUid fromEmail body = do
...
+        else setTeamInvitationTimeout <$> view settings
+    code <- liftIO $ DB.mkInvitationCode
+    mUrl <-
+      wrapClient $
+        if migratePersonalUser
+          then DB.mkInviteUrlPersonalUser showInvitationUrl tid code
+          else DB.mkInviteUrl showInvitationUrl tid code
```
Comment (fisx) [inline services/brig/src/Brig/Team/API.hs:277]:
```suggestion
        (if migratePersonalUser
          then DB.mkInviteUrlPersonalUser
          else DB.mkInviteUrl) showInvitationUrl tid code
```

File: services/brig/src/Brig/Team/API.hs [hunk 10]
```diff
@@ -239,20 +265,33 @@ createInvitation' tid mUid inviteeRole mbInviterUid fromEmail body = do
...
           mbInviterUid
           email
-          body.inviteeName
+          invRequest.inviteeName
           timeout
-    (newInv, code) <$ sendInvitationMail email tid fromEmail code body.locale
+          mUrl
+    if migratePersonalUser
+      then sendInvitationMailPersonalUser email tid fromEmail code invRequest.locale
+      else sendInvitationMail email tid fromEmail code invRequest.locale
```
Comment (fisx) [inline services/brig/src/Brig/Team/API.hs:293]:
```suggestion
    (if migratePersonalUser
      then sendInvitationMailPersonalUser
      else sendInvitationMail) email tid fromEmail code invRequest.locale
```

or even make the thing in parens a top-level function?

File: integration/test/Test/Teams.hs [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+          documents <- resp.json %. "documents" >>= asList
+          ids <- for documents ((%. "id") >=> asString)
+          ids `shouldContain` [ownerId]
+        refreshIndex domain
+        -- a team member can now search for the former personal user
+        bindResponse (searchContacts tm (user %. "name") domain) $ \resp -> do
+          resp.status `shouldMatchInt` 200
+          document <- resp.json %. "documents" >>= asList >>= assertOne
+          document %. "id" `shouldMatch` uid
+          document %. "team" `shouldMatch` tid
```
Comment (fisx) [inline integration/test/Test/Teams.hs:93]: now do the same thing with scim. does it work without us even thinking about that?  :)

(happy to do it in a separate PR if you prefer.)

File: services/brig/src/Brig/Team/API.hs [hunk 12]
```diff
@@ -355,3 +394,39 @@ changeTeamAccountStatuses tid s = do
...
+    Member (Embed HttpClientIO) r,
+    Member NotificationSubsystem r
+  ) =>
+  Local UserId ->
+  AcceptTeamInvitation ->
+  (Handler r) ()
+acceptTeamInvitationByPersonalUser luid req = do
+  mSelfProfile <- lift $ liftSem $ getSelfProfile luid
+  let mek = mkEmailKey <$> (userEmail . selfUser =<< mSelfProfile)
+  mInv <- API.findTeamInvitation mek req.code !>> toInvitationError
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/brig/src/Brig/Team/API.hs:414]: for what reason? i feel the names are good enough?

File: services/brig/src/Brig/Team/Email.hs [hunk 13]
```diff
@@ -56,6 +54,13 @@ sendMemberWelcomeMail to tid teamName loc = do
   let mail = MemberWelcomeEmail to tid teamName
   liftSem $ sendMail $ renderMemberWelcomeMail mail tpl branding
 
+sendInvitationMailPersonalUser :: (Member EmailSending r) => EmailAddress -> TeamId -> EmailAddress -> InvitationCode -> Maybe Locale -> (AppT r) ()
```
Comment (fisx) [inline services/brig/src/Brig/Team/Email.hs:57]: how many more instances of this pattern before it's time to write a generic `sendSomeMail` function?  :)

**Review by fisx — COMMENTED** [review]
i can't think of anything that we may have missed, but i'm not entirely confident that we haven't...  :thinking:

lots of comments, most of them do not require action.

File: integration/test/Test/Teams.hs (hunk 11 above)
Comment (fisx) [reply integration/test/Test/Teams.hs:93]: separate PR!

File: integration/test/Test/Teams.hs (hunk 3 above)
Comment (battermann) [reply integration/test/Test/Teams.hs:47]: I think this is not really related to the feature, but would only make the test nicer in this case.

**Review by fisx — CHANGES_REQUESTED** [review]
https://wearezeta.atlassian.net/browse/WPB-10658?focusedCommentId=128473

**Review by fisx — APPROVED** [review]

**Comment by fisx** [comment]
hints for merge conflict resolution with https://github.com/wireapp/wire-server/pull/4218: 

- `Brig.Team.DB` had been moved to `Wire.InvitationCodeStore` by them. `git diff origin/develop..WPB-10658-invitation-and-acceptance-of-individual-users-to-teams services/brig/src/Brig/Team/DB.hs` shows the changes we need to make happen there.
- `findTeamInvitation` (local function) in `Brig.API.User` has changed by them and has been moved to the top-level by us.
- `createInvitation'` in `Brig.Team.API` has been changed by both them and us.

All other conflicts should be trivial.

us:

```
git diff $(git merge-base origin/develop origin/WPB-10658-invitation-and-acceptance-of-individual-users-to-teams)..origin/WPB-10658-invitation-and-acceptance-of-individual-users-to-teams
```

them:

```
git diff $(git merge-base origin/develop origin/mangoiv/wpb-8887)..origin/mangoiv/wpb-8887
```

**Comment by fisx** [comment]
backup branch before rebase: WPB-10658-invitation-and-acceptance-of-individual-users-to-teams-bakleif

**Comment by fisx** [comment]
there are some conflicts left from the rebase, just follow the compiler errors.

File: services/brig/src/Brig/App.hs [hunk 14]
```diff
@@ -441,6 +442,10 @@ providerTemplates l = forLocale l <$> view provTemplates
 teamTemplates :: (MonadReader Env m) => Maybe Locale -> m (Locale, TeamTemplates)
 teamTemplates l = forLocale l <$> view tmTemplates
 
+-- this works because team templates is not affected by `forLocale`
```
Comment (fisx) [inline services/brig/src/Brig/App.hs:445]:
```suggestion
-- this works because team templates is not affected by `forLocale`; it is useful where we
-- use the `TeamTemplates` only for finding invitation url templates (those are not localized).
```

File: services/brig/src/Brig/Team/API.hs [hunk 15]
```diff
@@ -475,3 +561,53 @@ changeTeamAccountStatuses tid s = do
...
+    Member TinyLog r,
+    Member (Embed HttpClientIO) r,
+    Member NotificationSubsystem r,
+    Member InvitationCodeStore r,
+    Member PasswordStore r
+  ) =>
+  Local UserId ->
+  AcceptTeamInvitation ->
+  (Handler r) ()
+acceptTeamInvitationByPersonalUser luid req = do
```
Comment (fisx) [inline services/brig/src/Brig/Team/API.hs:581]: maybe this function could be reorganized to be a little clearer?

**Review by fisx — APPROVED** [review]
i'd like to look into `acceptTeamInvitationByPersonalUser`, but if you don't want to do that here that's fine, too.  :)
